src/coursescheduler/constraints.py

```diff
@@ -1,2 +1,164 @@
-from models import *
-from csp import Constraint
\ No newline at end of file
+from typing import List, TypeVar
+
+from .datamodels import temp_profs, temp_courses, courses, professors
+from .csp import Constraint
+
+V = TypeVar('V')  # variable type
+D = TypeVar('D')  # domain type
+
+
+# Hard constraint: instructors may only be assigned to courses for which they are qualified.
+# Note: this constraint currently isn't used.
+# Instead, we exclude any unqualified professors from the domain of each variable in CSP 1.
+class Qualified_Course_Prof(Constraint):
+    def __init__(self, course):
+        super().__init__([course])
+
+    def satisfied(self, assignment) -> bool:
+        course = self.variables[0]
+        if course not in assignment:
+            return True
+
+        prof = assignment[course]
+        if course in temp_profs[prof]["qualifiedCoursePreferences"]:
+            return True
+
+        return False
+
+class Course_Requires_PENG(Constraint):
+    def __init__(self, course):
+        super().__init__([course])
+
+    def satisfied(self, assignment) -> bool:
+        course = self.variables[0]
+        if course not in assignment:
+            return True
+
+        prof = assignment[course]
+        if temp_profs[prof]["isPeng"]:
+            return True
+
+        return False
+
+class Professor_Teaching_Load(Constraint):
+    def __init__(self, courses) -> None:
+        super().__init__(courses)
+
+    def satisfied(self, assignment) -> bool:
+        teaching_loads_dict = {prof : 0 for prof in temp_profs}
+
+        for course in self.variables:
+            if course not in assignment:
+                continue
+            prof = assignment[course]
+            teaching_loads_dict[prof] += 1
+
+        for prof, teachingLoad in teaching_loads_dict.items():
+            if teachingLoad > temp_profs[prof]["teachingObligations"]:
+                return False
+
+        return True
```

Comment:
I think this is OK. We would have to consider whether maintaining a tally across calls to satisfied() would cause problems when we backtrack (I think we would have to decrement the counts). I think we should apply this optimization only if we have time.